### PR TITLE
feat(mongo): add support for mongodb text indexes

### LINKED ISSUE #479: Question about Text Indexes
Are MongoDB text indexes supported? Something like this:
```javascript
db.reviews.createIndex( { comments: "text" } )
```

### PR DESCRIPTION
**Pull request by osayfun**
Added support for text indexes in MongoDB.
Added text index test.

Closes #479

### CODE REVIEW

File: lib/drivers/MongoDriver.ts [hunk 1]
```diff
@@ -104,14 +104,35 @@ export class MongoDriver extends DatabaseDriver<MongoConnection> {
...
+      if (index.type === 'text') {
+        const spec: {[key: string]: 'text'} = {};
+        properties.forEach(prop => spec[prop] = 'text');
+        fieldOrSpec = spec;
+      } else {
+        fieldOrSpec = properties;
+      }
+      promises.push(this.getConnection('write').getCollection(meta.name).createIndex(fieldOrSpec, {
+        name: index.name,
+        unique: index.type === 'uniques',
```
Comment (B4nan) [inline lib/drivers/MongoDriver.ts:121]: this is wrong, `index.type` is not the same as the parameter `type` you just removed


```suggestion
        unique: false,
```

File: tests/EntityManager.mongo.test.ts [hunk 2]
```diff
@@ -497,11 +497,11 @@ describe('EntityManagerMongo', () => {
       born_1: { key: { born: 1 }, name: 'born_1' },
       custom_idx_1: { key: { email: 1, name: 1 }, name: 'custom_idx_1' },
       age_uniq: { key: { age: 1 }, name: 'age_uniq', unique: true, partialFilterExpression: { age: { $exists: true } } },
-      email_1: { key: { email: 1 }, name: 'email_1', unique: true },
```
Comment (B4nan) [inline tests/EntityManager.mongo.test.ts:500]: this should stay there, I guess you broke something? or why did you remove it from the assertion? I guess it is still passing as we use `toMatchObject` here, but we need to make sure the index is defined there.

File: tests/entities/Book.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { BaseEntity3 } from './BaseEntity3';
 
 @Entity({ tableName: 'books-table' })
 @Unique({ properties: ['title', 'author'] })
+@Index({ properties: 'title', type: 'text'})
```
Comment (B4nan) [inline tests/entities/Book.ts:10]:
```suggestion
@Index({ properties: 'title', type: 'text' })
```

File: lib/drivers/MongoDriver.ts [hunk 4]
```diff
@@ -104,14 +104,35 @@ export class MongoDriver extends DatabaseDriver<MongoConnection> {
...
   }
 
-  private createIndexes(meta: EntityMetadata, type: 'indexes' | 'uniques') {
+  private createIndexes(meta: EntityMetadata) {
     const promises: Promise<string>[] = [];
+    meta.indexes.forEach(index => {
+      let fieldOrSpec: string | any;
+      const properties = Utils.flatten(Utils.asArray(index.properties).map(prop => meta.properties[prop].fieldNames));
+      if (index.type === 'text') {
+        const spec: {[key: string]: 'text'} = {};
```
Comment (B4nan) [inline lib/drivers/MongoDriver.ts:113]:
```suggestion
        const spec: Dictionary<string> = {};
```

File: lib/drivers/MongoDriver.ts [hunk 5]
```diff
@@ -104,14 +104,35 @@ export class MongoDriver extends DatabaseDriver<MongoConnection> {
     await Promise.all(promises);
   }
 
-  private createIndexes(meta: EntityMetadata, type: 'indexes' | 'uniques') {
+  private createIndexes(meta: EntityMetadata) {
     const promises: Promise<string>[] = [];
+    meta.indexes.forEach(index => {
+      let fieldOrSpec: string | any;
+      const properties = Utils.flatten(Utils.asArray(index.properties).map(prop => meta.properties[prop].fieldNames));
+      if (index.type === 'text') {
```
Comment (B4nan) [inline lib/drivers/MongoDriver.ts:112]: add blank line before control statement

File: lib/drivers/MongoDriver.ts [hunk 6]
```diff
@@ -104,14 +104,35 @@ export class MongoDriver extends DatabaseDriver<MongoConnection> {
...
+        const spec: {[key: string]: 'text'} = {};
+        properties.forEach(prop => spec[prop] = 'text');
+        fieldOrSpec = spec;
+      } else {
+        fieldOrSpec = properties;
+      }
```
Comment (B4nan) [inline lib/drivers/MongoDriver.ts:118]: add blank line after control statement

**Comment by B4nan** [comment]
Would be also good to mention this in the docs here: https://mikro-orm.io/docs/usage-with-mongo#indexes

File: tests/EntityManager.mongo.test.ts (hunk 2 above)
Comment (osayfun) [reply tests/EntityManager.mongo.test.ts:500]: Oh sorry, my bad! I was just playing around!

**Comment by osayfun** [comment]
MongoDB has different indexes. I think it would be better to have a section for them in docs.

**Comment by B4nan** [comment]
The page I linked is about differences with mongo driver and about index support in particular, so I think it is the right place (at least with the current docs structure). Feel free to add the content to other places too (or create new section about it), but at least a brief note should be part of the linked page too.

File: lib/drivers/MongoDriver.ts [hunk 7]
```diff
@@ -104,14 +104,37 @@ export class MongoDriver extends DatabaseDriver<MongoConnection> {
...
+    meta.indexes.forEach(index => {
+      let fieldOrSpec: string | any;
+      const properties = Utils.flatten(Utils.asArray(index.properties).map(prop => meta.properties[prop].fieldNames));
+
+      if (index.type === 'text') {
+        const spec: Dictionary<string> = {};
+        properties.forEach(prop => spec[prop] = 'text');
+        fieldOrSpec = spec;
+      } else {
+
```
Comment (B4nan) [inline lib/drivers/MongoDriver.ts:118]:
```suggestion
```

the blank line should be after the whole block (where i put the comment, so line 121)

**Comment by B4nan** [comment]
Perfect, looking good, thanks again. Let's wait for green CI and we can merge. Will ship this as a patch release as its quite minor thing, probably tomorrow, want to fix 2 small bugs first.

**Comment by osayfun** [comment]
> Perfect, looking good, thanks again. Let's wait for green CI and we can merge. Will ship this as a patch release as its quite minor thing, probably tomorrow, want to fix 2 small bugs first.

It's my pleasure. Thanks for the great package.
